File: src/widget/widgetLogQueue.ts
/**
 * Taps on the Log Today widget, on their way to the journal.
 *
 * ── Why a queue, and not a write ──────────────────────────────────────
 *
 * The journal is an ENCRYPTED blob. A widget's broadcast receiver is
 * Kotlin, it has no access to the key, and even if it did it would have to
 * reimplement `upsertEntry`'s rule that an existing status is never
 * overwritten — so an accidental double press cannot clear a "missed" the
 * user set on purpose. Two implementations of that rule is one too many;
 * the one in `prayerLogAction.logPrayerOnTime` is the writer, and it stays
 * the only one.
 *
 * The alternative was a HeadlessJsTaskService: spin up the whole React
 * Native runtime on a tap, run the write in JS, tear it down. That is
 * seconds of work for one line of journal, Android 12+ restricts starting
 * services from the background anyway, and the user is standing there
 * waiting for a checkmark. So the receiver appends to a queue in the same
 * SharedPreferences the payload already lives in — a few bytes, instant —
 * and the app drains it through the real writer at the next opportunity.
 *
 * ── What the user sees ────────────────────────────────────────────────
 *
 * Immediately: the tick. The widget renders the queue over the payload, so
 * a tapped prayer reads as logged the moment it is tapped, whether or not
 * the app has run since.
 *
 * ── The undo window falls out of this for free ────────────────────────
 *
 * Tapping the same prayer again inside the undo window removes it from the
 * queue instead of adding a second entry. Nothing had been written yet, so
 * "undo" is a deletion from a list rather than a journal edit that has to
 * decide what the previous status was.
 *
 * ── Which day an entry lands on ───────────────────────────────────────
 *
 * The date is stamped when the tap happens and travels with the entry, for
 * the reason `prayerLogAction` gives about notifications: Isha fires at
 * 23:40 in a Swedish winter and is answered at 00:05, and a drain that used
 * "today" would credit the wrong day and leave the real one blank.
 */
import type { JournalPrayer } from '../journal/journal';

/** One queued tap. Short keys — this rides in a SharedPreferences string. */
export type WidgetLogEntry = {
  /** Local YYYY-MM-DD the tap was FOR, stamped at tap time. */
  d: string;
  /** Which prayer. */
  p: JournalPrayer;
  /** Epoch ms of the tap, for the undo window and for ordering. */
  t: number;
};

/** How long a tap can be taken back by tapping the same prayer again. */
export const UNDO_WINDOW_MS = 60_000;

/**
 * How stale a queued tap may be and still be written.
 *
 * A queue that has not drained in a fortnight means the app has not been
 * opened in a fortnight, and writing those entries then would silently
 * backfill two weeks of history the user has had no chance to review. The
 * journal is a record of what someone did; it should not gain a fortnight
 * in one go without them seeing it happen.
 */
export const MAX_QUEUE_AGE_MS = 14 * 24 * 60 * 60 * 1000;

const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;
const PRAYERS: readonly JournalPrayer[] = [
  'Fajr',
  'Dhuhr',
  'Asr',
  'Maghrib',
  'Isha',
];

/**
 * Read a stored queue back, keeping only entries that could have come from
 * a real tap.
 *
 * The same discipline as every other blob this app reads: the string comes
 * from another process, and a malformed entry is dropped rather than
 * written into someone's record of their own worship.
 */
export function coerceLogQueue(input: unknown): WidgetLogEntry[] {
  if (!Array.isArray(input)) return [];
  const out: WidgetLogEntry[] = [];
  for (const item of input) {
    if (!item || typeof item !== 'object') continue;
    const r = item as Record<string, unknown>;
    if (typeof r.d !== 'string' || !DATE_RE.test(r.d)) continue;
    if (typeof r.p !== 'string' || !PRAYERS.includes(r.p as JournalPrayer)) {
      continue;
    }
    const t = typeof r.t === 'number' && Number.isFinite(r.t) ? r.t : 0;
    if (t <= 0) continue;
    out.push({ d: r.d, p: r.p as JournalPrayer, t });
  }
  return out;
}

/**
 * Apply one tap to a queue.
 *
 * Exported and pure so the rule can be tested without a device, and so the
 * Kotlin side has something to be checked AGAINST rather than a second
 * opinion — the receiver implements the same three cases and the tests
 * here are what say what they should be.
 *
 * The three cases:
 *   • Already queued, inside the undo window → remove it. This is the undo.
 *   • Already queued, outside the window → leave it. A second tap an hour
 *     later is someone confirming, not someone retracting.
 *   • Not queued → add it.
 */
export function applyTap(
  queue: WidgetLogEntry[],
  date: string,
  prayer: JournalPrayer,
  now: number,
  undoWindowMs: number = UNDO_WINDOW_MS,
): WidgetLogEntry[] {
  const existing = queue.find(e => e.d === date && e.p === prayer);
  if (existing) {
    if (now - existing.t <= undoWindowMs) {
      return queue.filter(e => !(e.d === date && e.p === prayer));
    }
    return queue;
  }
  return [...queue, { d: date, p: prayer, t: now }];
}

/**
 * Split a queue into what should be written now and what should be dropped.
 *
 * Dropped entries are NOT written and not kept: a tap older than the window
 * has already been superseded by the user opening the app and seeing the
 * day as it really is.
 */
export function partitionQueue(
  queue: WidgetLogEntry[],
  now: number,
  maxAgeMs: number = MAX_QUEUE_AGE_MS,
): { write: WidgetLogEntry[]; stale: WidgetLogEntry[] } {
  const write: WidgetLogEntry[] = [];
  const stale: WidgetLogEntry[] = [];
  for (const e of queue) {
    (now - e.t > maxAgeMs ? stale : write).push(e);
  }
  // Oldest first, so a day's five prayers land in the order they were
  // tapped rather than in whatever order the array happened to hold.
  write.sort((a, b) => a.t - b.t);
  return { write, stale };
}

/** Everything a drain needs, injected so the whole thing is testable. */
export type DrainDeps = {
  /** Read the queue AND clear it, atomically, on the native side. */
  take: () => Promise<unknown>;
  /** The real writer — `logPrayerOnTime`. */
  write: (date: string, prayer: JournalPrayer, now?: Date) => Promise<boolean>;
  now?: number;
};

export type DrainResult = {
  written: number;
  /** Taps that named a prayer already carrying a status. Not an error: the
   *  user logged it in the app before the queue drained. */
  skipped: number;
  dropped: number;
  failed: number;
};

/**
 * Write every queued tap through the real journal writer.
 *
 * Failures are counted rather than thrown. A drain runs on app foreground
 * and from the notification background handler, and neither of those is a
 * place to surface an exception — the queue was already taken, so a throw
 * here would lose the taps rather than report them.
 */
export async function drainWidgetLogQueue(
  deps: DrainDeps,
): Promise<DrainResult> {
  const now = deps.now ?? Date.now();
  const raw = await deps.take().catch(() => null);
  const queue = coerceLogQueue(raw);
  if (queue.length === 0) {
    return { written: 0, skipped: 0, dropped: 0, failed: 0 };
  }
  const { write, stale } = partitionQueue(queue, now);
  let written = 0;
  let skipped = 0;
  let failed = 0;
  for (const e of write) {
    try {
      const didWrite = await deps.write(e.d, e.p, new Date(e.t));
      if (didWrite) written += 1;
      else skipped += 1;
    } catch {
      failed += 1;
    }
  }
  return { written, skipped, dropped: stale.length, failed };
}

File: android/app/src/main/java/com/mubarok/ibadat365/WidgetLogQueue.kt
package com.mubarok.ibadat365

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Taps on the Log Today widget, waiting for the app to write them.
 *
 * The journal is an encrypted blob whose key lives on the JS side, so this
 * receiver cannot write it — and should not want to. `logPrayerOnTime` in
 * `prayerLogAction.ts` already owns that write, including the rule that an
 * existing status is never overwritten, and a second implementation of that
 * rule in Kotlin is one that can drift from the first.
 *
 * So a tap lands here, in the same SharedPreferences the payload already
 * uses, and the app drains it through the real writer when it next runs.
 * The widget renders this queue over the payload, so the checkmark appears
 * on the tap rather than on the next launch.
 *
 * THE RULES BELOW MUST MATCH `src/widget/widgetLogQueue.ts`. That file's
 * tests are what say what they are; this is the second implementation of a
 * three-case rule and the only defence against the two disagreeing is that
 * the cases are few enough to state in one place each.
 */
object WidgetLogQueue {
  const val PREFS_QUEUE_KEY = "widget_log_queue"

  /** Tapping the same prayer again within this takes the tap back. */
  const val UNDO_WINDOW_MS = 60_000L

  /** The five that can be logged, in the order they are prayed. */
  val PRAYERS = listOf("Fajr", "Dhuhr", "Asr", "Maghrib", "Isha")

  data class Entry(val date: String, val prayer: String, val at: Long)

  private val DATE_RE = Regex("^\\d{4}-\\d{2}-\\d{2}$")

  private fun prefs(context: Context) =
    context.getSharedPreferences(PrayerWidgetProvider.PREFS_NAME, Context.MODE_PRIVATE)

  /**
   * Read the queue back, dropping anything that could not have come from a
   * real tap — the same discipline the JS side applies, for the same
   * reason: this string is written by another process and its contents end
   * up in someone's record of their own worship.
   */
  fun read(context: Context): List<Entry> {
    val raw = prefs(context).getString(PREFS_QUEUE_KEY, null) ?: return emptyList()
    return parse(raw)
  }

  fun parse(raw: String): List<Entry> {
    val out = mutableListOf<Entry>()
    try {
      val arr = JSONArray(raw)
      for (i in 0 until arr.length()) {
        val o = arr.optJSONObject(i) ?: continue
        val d = o.optString("d")
        val p = o.optString("p")
        val t = o.optLong("t", 0L)
        if (!DATE_RE.matches(d)) continue
        if (!PRAYERS.contains(p)) continue
        if (t <= 0L) continue
        out.add(Entry(d, p, t))
      }
    } catch (_: Exception) {
      return emptyList()
    }
    return out
  }

  fun serialize(entries: List<Entry>): String {
    val arr = JSONArray()
    for (e in entries) {
      arr.put(JSONObject().put("d", e.date).put("p", e.prayer).put("t", e.at))
    }
    return arr.toString()
  }

  /**
   * Apply one tap. Mirrors `applyTap` in widgetLogQueue.ts:
   *
   *   • queued, inside the undo window  → remove it (this is the undo)
   *   • queued, outside the window      → leave it; a tap an hour later is
   *                                       someone confirming, and silently
   *                                       un-logging a prayer they believe
   *                                       they logged is the worst outcome
   *                                       available here
   *   • not queued                      → add it
   *
   * Returns the new queue so the caller can tell whether anything changed.
   */
  fun applyTap(
    queue: List<Entry>,
    date: String,
    prayer: String,
    now: Long,
    undoWindowMs: Long = UNDO_WINDOW_MS,
  ): List<Entry> {
    val existing = queue.firstOrNull { it.date == date && it.prayer == prayer }
    if (existing != null) {
      return if (now - existing.at <= undoWindowMs) {
        queue.filterNot { it.date == date && it.prayer == prayer }
      } else {
        queue
      }
    }
    return queue + Entry(date, prayer, now)
  }

  /** Record a tap and persist it. */
  fun tap(context: Context, date: String, prayer: String, now: Long = System.currentTimeMillis()) {
    val next = applyTap(read(context), date, prayer, now)
    prefs(context).edit().putString(PREFS_QUEUE_KEY, serialize(next)).apply()
  }

  /**
   * Hand the queue over and clear it in one step.
   *
   * One step because the app is about to write these to the journal, and a
   * read-then-clear that is interrupted between the two either loses taps
   * or writes them twice. `apply()` is asynchronous but the in-memory value
   * changes immediately, so a second caller sees the cleared queue.
   */
  fun take(context: Context): List<Entry> {
    val entries = read(context)
    if (entries.isNotEmpty()) {
      prefs(context).edit().remove(PREFS_QUEUE_KEY).apply()
    }
    return entries
  }

  /** Which prayers are queued for `date` — what the widget draws as ticked. */
  fun pendingFor(context: Context, date: String): Set<String> =
    read(context).filter { it.date == date }.map { it.prayer }.toSet()
}
